nilearn/decoding/decoder.py

```diff
@@ -106,41 +106,130 @@ def _check_param_grid(estimator, X, y, param_grid=None):
+        the affine is considered the same for all.
+
+    y: array or list of shape (n_samples)
+        The dependent variable (age, sex, IQ, yes/no, etc.).
+        Target variable to predict. Must have exactly as many elements as
+        3D images in niimg.
+
+    Returns
+    -------
+    param_grid: dict of str to sequence, or sequence of such. Sensible default
+    dict has size 1 for linear models.
+    """
+    param_grid = {}
+
+    # define loss function
+    if isinstance(estimator, LogisticRegression):
+        loss = "log"
+    elif isinstance(estimator, (LinearSVC, RidgeCV, RidgeClassifierCV, SVR)):
+        loss = "squared_hinge"
+    elif isinstance(estimator, (DummyClassifier, DummyRegressor)):
+        if estimator.strategy in ["constant"]:
+            message = (
+                "Dummy classification implemented only for strategies"
+                ' "most_frequent", "prior", "stratified"'
+            )
+            raise NotImplementedError(message)
+    else:
+        raise ValueError(
+            "Invalid estimator. The supported estimators are:"
+            f" {list(SUPPORTED_ESTIMATORS.keys())}"
+        )
+    # define sensible default for different types of estimators
+    if hasattr(estimator, "penalty") and (estimator.penalty == "l1"):
+        min_c = l1_min_c(X, y, loss=loss)
+    else:
+        min_c = 0.5
+
+    if isinstance(estimator, (RidgeCV, RidgeClassifierCV)):
+        param_grid["alphas"] = [np.geomspace(1e-3, 1e4, 8)]
+    elif not isinstance(estimator, (DummyClassifier, DummyRegressor)):
+        param_grid["C"] = np.array([2, 20, 200]) * min_c
+    else:
         param_grid = {}
-        # define loss function
-        if isinstance(estimator, LogisticRegression):
-            loss = "log"
-        elif isinstance(
-            estimator, (LinearSVC, RidgeCV, RidgeClassifierCV, SVR)
+
+    return param_grid
+
+
+def _wrap_param_grid(param_grid, param_name):
+    """Wrap a parameter's sequence of values with an outer list.
+
+    This can be
+    desirable for models featuring built-in cross-validation, as it would leave
+    it to the model's internal (optimized) cross-validation to loop over
+    hyperparameter values. Does nothing if the parameter is already wrapped.
+
+    Parameters
+    ----------
+    param_grid : dict of str to sequence, or sequence of such
+        The parameter grid to wrap, as a dictionary mapping estimator
+        parameters to sequences of allowed values.
+    param_name : str
+        Name of parameter whose sequence of values should be wrapped
+
+    Returns
+    -------
+    param_grid_wrapped: dict of str to sequence, or sequence of such
+        The updated parameter grid
+    """
+    if param_grid is None:
+        return param_grid
+
+    # param_grid can be either a dict or a sequence of dicts
+    # we make sure that it is a sequence we can loop over
+    input_is_dict = isinstance(param_grid, dict)
```

Comment:
perhaps `collections.abc.Mapping` rather than `dict`?

Reply:
I am not sure this is a good idea because `param_grid` is eventually (in `_parallel_fit()`) passed to `sklearn.model_selection.ParameterGrid`, which explicitly checks that each grid is a `dict` (see [here](https://github.com/scikit-learn/scikit-learn/blob/364c77e04/sklearn/model_selection/_search.py#L106)). If I pass a `Mapping` object that is not a `dict` I get a `TypeError` when instantiating `ParameterGrid`:

```python
from collections.abc import Mapping
from sklearn.model_selection import ParameterGrid

class NonDictMapping(Mapping):
    def __init__(self, data=None) -> None:
        if data is None:
            data = {}
        self.data = data

    def __getitem__(self, __key):
        return self.data[__key]
    
    def __iter__(self):
        return self.data.__iter__()
    
    def __len__(self) -> int:
        return self.data.__len__()

param_grid = NonDictMapping()
print(ParameterGrid(param_grid))
``` 

I guess we could accept `Mapping` objects and internally cast them to `dict`s, but not sure if it's worth it. What do you think @jeromedockes?

Reply:
good point -- sorry I missed that! in this case we can keep the check as it is